### PR TITLE
Dungeon Secrets

### PR DESCRIPTION
**Pull request by kevinthegreat1**
Adds dungeon secrets waypoints.  
This is currently usable but needs more testing, which is why I'm opening a pr now.  

## New changes
* Fixed windows crash.  
* Fixed dungeon start detection (the player no longer needs to stay in the entrance room for a few seconds after dungeon start).  
* Added ability to disable certain types of waypoints.  
* Added ability to mark secrets as found or missing.  
* Fixed getMapEntrancePosAndRoomSize crash.  
* Turns off a lever waypoint when the lever is interacted with.  
* Resets on a new connection/world.  
* Double check 10 blocks after matching a room to make sure that the room is correct.  

## Known issues
* There may be a small chance of rooms not getting detected.  

## Improvements
* The ability to add custom waypoints.

### CODE REVIEW

**Comment by kevinthegreat1** [comment]
Meant to open as draft.

File: src/main/java/me/xmrvizzy/skyblocker/skyblock/dungeon/secrets/DungeonMapUtils.java [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        for (MapIcon icon : map.getIcons()) {
+            if (icon.getType() == MapIcon.Type.FRAME) {
+                return new Vector2i((icon.getX() >> 1) + 64, (icon.getZ() >> 1) + 64);
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    public static Pair<Vector2ic, Integer> getMapEntrancePosAndRoomSize(@NotNull MapState map) {
```
Comment (AzureAaron) [inline src/main/java/me/xmrvizzy/skyblocker/skyblock/dungeon/secrets/DungeonMapUtils.java:55]: Should use ObjectIntPair

File: src/main/java/me/xmrvizzy/skyblocker/skyblock/dungeon/secrets/DungeonMapUtils.java [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+            }
+            pos = new Vector2i(mapPos).add(0, 10);
+            if (checked.add(pos)) {
+                posToCheck.add(pos);
+            }
+        }
+        return null;
+    }
+
+    private static Pair<Vector2ic, Integer> getMapEntrancePosAndRoomSizeAt(MapState map, Vector2ic mapPosImmutable) {
```
Comment (AzureAaron) [inline src/main/java/me/xmrvizzy/skyblocker/skyblock/dungeon/secrets/DungeonMapUtils.java:88]: Should use ObjectIntPair

File: src/main/java/me/xmrvizzy/skyblocker/skyblock/dungeon/secrets/DungeonSecrets.java [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+        ItemStack stack = player.getInventory().main.get(8);
+        if (!stack.isOf(Items.FILLED_MAP)) {
+            return;
+        }
+        MapState map = FilledMapItem.getMapState(FilledMapItem.getMapId(stack), client.world);
+        if (map == null) {
+            return;
+        }
+        if (mapEntrancePos == null || mapRoomSize == 0) {
+            Pair<Vector2ic, Integer> mapEntrancePosAndSize = DungeonMapUtils.getMapEntrancePosAndRoomSize(map);
```
Comment (AzureAaron) [inline src/main/java/me/xmrvizzy/skyblocker/skyblock/dungeon/secrets/DungeonSecrets.java:241]: Should be ObjectIntPair

**Review by AzureAaron — CHANGES_REQUESTED** [review]

**Review by AzureAaron — APPROVED** [review]

**Review by LifeIsAParadox — APPROVED** [review]
lgtm

**Pull request merged**